### PR TITLE
Feature： rgw callback implement

### PR DESCRIPTION
**Pull request by joke-lee**
callback feature like aliyun oss
https://www.alibabacloud.com/help/doc-detail/50092.html?spm=a2c5t.11065259.1996646101.searchclickresult.5970356bJxLX41

https://www.alibabacloud.com/help/doc-detail/31989.html?spm=a2c5t.11065259.1996646101.searchclickresult.5970356bJxLX41

https://www.alibabacloud.com/help/doc-detail/31922.html?spm=a2c5t.11065259.1996646101.searchclickresult.5970356bJxLX41

## generate pem file for sign
```
openssl genrsa -out /ceph/myPrivateKey.pem  2048
openssl rsa -pubout -in /ceph/myPrivateKey.pem -out /ceph/myPublicKey.pem
```
## upload myPublicKey.pem
```
s3cmd mb s3://security
s3cmd put  /ceph/myPublicKey.pem  -P  s3://security
```
modify ceph.conf and restart rgw
```
rgw_callback_private_rsa_key = /ceph/myPrivateKey.pem
rgw_callback_public_rsa_key_url = "http://127.0.0.1/security/myPublicKey.pem"
rgw_callback_max_response = 1
rgw_callback_callback_var_max_length = 5
```

## callback server script
callbackserver.py
```
import socket
import httplib
import base64
import md5
import urllib2
from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
from M2Crypto import RSA
from M2Crypto import BIO

def get_local_ip():
    try:
        csock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        csock.connect(('8.8.8.8', 80))
        (addr, port) = csock.getsockname()
        csock.close()
        return addr
    except socket.error:
        return ""

class MyHTTPRequestHandler(BaseHTTPRequestHandler):

    def do_POST(self):
        pub_key_url = ''
        try:
            pub_key_url_base64 = self.headers['x-callback-pub-key-url']
            pub_key_url = pub_key_url_base64.decode('base64')
            url_reader = urllib2.urlopen(pub_key_url)
            pub_key = url_reader.read()
        except:
            print 'pub_key_url : ' + pub_key_url
            print 'Get pub key failed!'
            self.send_response(400)
            self.end_headers()
            return

        #get authorization
        authorization_base64 = self.headers['authorization']
        authorization = authorization_base64.decode('base64')

        #get callback body
        content_length = self.headers['content-length']
        callback_body = self.rfile.read(int(content_length))

        #compose authorization string
        auth_str = ''
        pos = self.path.find('?')
        if -1 == pos:
            auth_str = self.path + '\n' + callback_body
        else:
            auth_str = urllib2.unquote(self.path[0:pos]) + self.path[pos:] + '\n' + callback_body
        print auth_str

        #verify authorization
        auth_md5 = md5.new(auth_str).digest()
        bio = BIO.MemoryBuffer(pub_key)
        rsa_pub = RSA.load_pub_key_bio(bio)
        try:
            result = rsa_pub.verify(auth_md5, authorization, 'md5')
        except :
            result = False

        if not result:
            print 'Authorization verify failed!'
            print 'Public key : %s' % (pub_key)
            print 'Auth string : %s' % (auth_str)
            self.send_response(400)
            self.end_headers()
            return

        resp_body = '{"Status":"OK"}' 
        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Content-Length', str(len(resp_body)))
        self.end_headers()
        self.wfile.write(resp_body)

class MyHTTPServer(HTTPServer):
    def __init__(self, host, port):
        HTTPServer.__init__(self, (host, port), MyHTTPRequestHandler)


if '__main__' == __name__:
    server_ip = "127.0.0.1"
    server_port = 23450
    server = MyHTTPServer(server_ip, server_port)
    server.serve_forever()
```

## putobject.py script
```
# -*- coding: UTF-8 -*-
import requests
import hmac
from hashlib import sha1 as sha
py3k = False
try:
    from urlparse import urlparse, unquote
    from base64 import encodestring
except:
    py3k = True
    from urllib.parse import urlparse, unquote
    from base64 import encodebytes as encodestring

from email.utils import formatdate
from requests.auth import AuthBase
class S3Auth(AuthBase):
    """Attaches AWS Authentication to the given Request object."""

    service_base_url = 's3.amazonaws.com'
    # List of Query String Arguments of Interest
    special_params = [
        'acl', 'location', 'logging', 'partNumber', 'policy', 'requestPayment',
        'torrent', 'versioning', 'versionId', 'versions', 'website', 'uploads',
        'uploadId', 'response-content-type', 'response-content-language',
        'response-expires', 'response-cache-control', 'delete', 'lifecycle',
        'response-content-disposition', 'response-content-encoding', 'tagging',
        'notification', 'cors', 'syncing'
    ]

    def __init__(self, access_key, secret_key, service_url=None):
        if service_url:
            self.service_base_url = service_url
        self.access_key = str(access_key)
        self.secret_key = str(secret_key)

    def __call__(self, r):
        # Create date header if it is not created yet.
        if 'date' not in r.headers and 'x-amz-date' not in r.headers:
            r.headers['date'] = formatdate(
                timeval=None,
                localtime=False,
                usegmt=True)
        signature = self.get_signature(r)
        if py3k:
            signature = signature.decode('utf-8')
        r.headers['Authorization'] = 'AWS %s:%s' % (self.access_key, signature)
        return r

    def get_signature(self, r):
        canonical_string = self.get_canonical_string(
            r.url, r.headers, r.method)
        if py3k:
            key = self.secret_key.encode('utf-8')
            msg = canonical_string.encode('utf-8')
        else:
            key = self.secret_key
            msg = canonical_string
        h = hmac.new(key, msg, digestmod=sha)
        return encodestring(h.digest()).strip()

    def get_canonical_string(self, url, headers, method):
        parsedurl = urlparse(url)
        objectkey = parsedurl.path[1:]
        query_args = sorted(parsedurl.query.split('&'))

        bucket = parsedurl.netloc[:-len(self.service_base_url)]
        if len(bucket) > 1:
            # remove last dot
            bucket = bucket[:-1]

        interesting_headers = {
            'content-md5': '',
            'content-type': '',
            'date': ''}
        for key in headers:
            lk = key.lower()
            try:
                lk = lk.decode('utf-8')
            except:
                pass
            if headers[key] and (lk in interesting_headers.keys()
                                 or lk.startswith('x-amz-')):
                interesting_headers[lk] = headers[key].strip()

        # If x-amz-date is used it supersedes the date header.
        if not py3k:
            if 'x-amz-date' in interesting_headers:
                interesting_headers['date'] = ''
        else:
            if 'x-amz-date' in interesting_headers:
                interesting_headers['date'] = ''

        buf = '%s\n' % method
        for key in sorted(interesting_headers.keys()):
            val = interesting_headers[key]
            if key.startswith('x-amz-'):
                buf += '%s:%s\n' % (key, val)
            else:
                buf += '%s\n' % val

        # append the bucket if it exists
        if bucket != '':
            buf += '/%s' % bucket

        # add the objectkey. even if it doesn't exist, add the slash
        buf += '/%s' % objectkey

        params_found = False

        # handle special query string arguments
        for q in query_args:
            k = q.split('=')[0]
            if k in self.special_params:
                buf += '&' if params_found else '?'
                params_found = True

                try:
                    k, v = q.split('=', 1)

                except ValueError:
                    buf += q
                else:
                    buf += '{key}={value}'.format(key=k, value=unquote(v))

        return buf

host = '127.0.0.1'
access_key = ''
secret_key = ''
bucketname = ''
objectname = ''

import base64
import json
import requests
import logging
from requests_toolbelt.utils import dump
logging.basicConfig(level=logging.DEBUG)

callback_dict = {}
callback_dict['callbackUrl'] = 'http://127.0.0.1:23450/index.php?id=1&index=2'
callback_dict['callbackBody'] = 'bucket=${bucket}&filename=${object}&size=${size}&mimeType=${mimeType}&my_var=${x:var1}'
callback_dict['callbackBodyType'] = 'application/x-www-form-urlencoded'
callback_param = json.dumps(callback_dict).strip()
base64_callback_body = base64.b64encode(callback_param)
headers={'x-amz-meta-callback': base64_callback_body}
callback_var = {}
callback_var["x:var1"]='value1'
callback_var["x:var2"]='value2'
callback_var_param = json.dumps(callback_var).strip()
base64_callback_var = base64.b64encode(callback_var_param)
headers['x-amz-meta-callback-var'] = base64_callback_var

url="http://%s/%s/%s" % (host, bucketname, objectname)
response = requests.put(url, auth=S3Auth(access_key, secret_key,service_url='127.0.0.1'),headers=headers, data="this is content")
data = dump.dump_all(response)
print(data.decode('utf-8'))
```

```
[root@localhost ceph]# python putcallback.py
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 192.168.153.181
DEBUG:urllib3.connectionpool:http://192.168.153.181:80 "PUT /bucket1/obj.text HTTP/1.1" 200 16
< PUT /bucket1/obj.text HTTP/1.1
< Host: 192.168.153.181
< Connection: keep-alive
< Accept-Encoding: gzip, deflate
< Accept: */*
< User-Agent: python-requests/2.18.4
< x-amz-meta-callback: eyJjYWxsYmFja0JvZHlUeXBlIjogImFwcGxpY2F0aW9uL3gtd3d3LWZvcm0tdXJsZW5jb2RlZCIsICJjYWxsYmFja0JvZHkiOiAiYnVja2V0PSR7YnVja2V0fSZmaWxlbmFtZT0ke29iamVjdH0mc2l6ZT0ke3NpemV9Jm1pbWVUeXBlPSR7bWltZVR5cGV9Jm15X3Zhcj0ke3g6dmFyMX0iLCAiY2FsbGJhY2tVcmwiOiAiaHR0cDovLzE5Mi4xNjguMTUzLjE4MToyMzQ1MC9pbmRleC5waHA/aWQ9MSZpbmRleD0yIn0=
< x-amz-meta-callback-var: eyJ4OnZhcjEiOiAidmFsdWUxIiwgIng6dmFyMiI6ICJ2YWx1ZTIifQ==
< Content-Length: 15
< date: Thu, 17 May 2018 07:29:51 GMT
< Authorization: AWS yly:nY8tlBPeWsxB9xCKCiKiowzza0M=
<
< this is content
> HTTP/1.1 200 OK
> Content-Length: 16
> Accept-Ranges: bytes
> x-amz-request-id: tx000000000000000000007-005afd2f6f-8552-default
> Content-Type: application/json
> Date: Thu, 17 May 2018 07:29:52 GMT
> Connection: Keep-Alive
>
{"Status":"OK"}

the callback server recieve the rgw request
[root@localhost ceph]# python callbackserver.py
/index.php?id=1&index=2
bucket=bucket1&filename=obj.text&size=15&mimeType=text%2Fplain&my_var=value1
192.168.153.181 - - [17/May/2018 03:29:52] "POST /index.php?id=1&index=2 HTTP/1.1" 200 -
```

### CODE REVIEW

**Comment by joke-lee** [comment]
hi, @cbodley do you mind to help review this ? thanks

**Comment by cbodley** [comment]
cc @yehudasa @mattbenjamin

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -1405,6 +1407,9 @@ static int get_success_retcode(int code)
 
 void RGWPutObj_ObjStore_S3::send_response()
 {
+  if(op_ret == 0 && process_callback(s, this, op_ret, etag, boost::none))
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_s3.cc]: this seems to require being conditional (efficiently for the common case where it's not happening)

File: src/rgw/rgw_rest_callback_s3.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+#include "rgw_rest.h"
+#include "rgw_rest_s3.h"
+
+#include "rgw/rgw_b64.h"
+#include "include/assert.h"
+
+#define dout_subsys ceph_subsys_rgw
+
+class RGWCallBackClient : public RGWPostHTTPData {
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_callback_s3.h]: this is an alibaba extension--seems fine to support it, but other protocols/behaviors might be useful too

**Pull request closed without merging**